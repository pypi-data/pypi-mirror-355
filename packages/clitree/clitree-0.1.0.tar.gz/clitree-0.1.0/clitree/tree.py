from typing import Any, Callable, Iterable, TypeVar

Node = Any
T = TypeVar("T")
NodeValueGetter = Callable[[Node], T]


def _get_value(node: Node, key: str | NodeValueGetter[T]) -> T | None:
    """
    Get the value of a key or attribute from a node.

    Args:
        node: The node to get the value from
        key: The key or attribute to get the value from. Can be a string key or a callable

    Returns:
        The value of the key or attribute
    """
    if callable(key):
        return key(node)
    if hasattr(node, "get"):
        return node.get(key)
    return getattr(node, key, None)


def _get_name(node: Node, name: str | NodeValueGetter[str]) -> str:
    result = _get_value(node, name)
    if result is None:
        raise ValueError(f"Node name resolved to None: {node}")
    return str(result)


def _get_children(
    node: Node, children: str | NodeValueGetter[Iterable[Node] | None]
) -> Iterable[Node] | None:
    result = _get_value(node, children)
    if result is None:
        return None
    return result


def _build_tree(
    node: Node,
    name: str | NodeValueGetter[str],
    children: str | NodeValueGetter[Iterable[Node] | None],
    prefix: str = "",
) -> str:
    node_name = _get_name(node, name)
    node_children = _get_children(node, children) or []
    children_list = list(node_children)

    result = [node_name]

    for i, child in enumerate(children_list):
        is_last_child = i == len(children_list) - 1
        connector = "└── " if is_last_child else "├── "
        child_prefix = prefix + ("    " if is_last_child else "│   ")
        result.append(
            prefix + connector + _build_tree(child, name, children, child_prefix)
        )

    return "\n".join(result)


def tree(
    data: Node,
    name: str | NodeValueGetter[str] = "name",
    children: str | NodeValueGetter[Iterable[Node] | None] = "children",
) -> str:
    """
    Generate a tree-like string representation of a hierarchical data structure.

    Args:
        data: The root node of the tree structure
        name: How to extract the name of a node. Can be a string key or a callable
        children: How to extract children of a node. Can be a string key or a callable

    Returns:
        A string representation of the tree structure

    Raises:
        ValueError: If the name resolves to None
        TypeError: If children are not iterable
    """
    return _build_tree(data, name, children)
