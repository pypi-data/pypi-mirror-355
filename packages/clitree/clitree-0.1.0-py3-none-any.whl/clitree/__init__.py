"""A library to draw tree structures like the CLI utility 'tree' does."""

__version__ = "0.1.0"

from .tree import tree

__all__ = ["tree"]
