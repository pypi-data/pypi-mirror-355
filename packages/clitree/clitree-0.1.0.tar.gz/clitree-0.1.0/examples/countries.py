from clitree import tree

# Nested dictionary structure for countries and capitals
COUNTRY_DATA = {
    "Europe": [
        {"country": "France", "capital": "Paris"},
        {"country": "Germany", "capital": "Berlin"},
        {"country": "Italy", "capital": "Rome"},
        {"country": "Spain", "capital": "Madrid"},
        {"country": "United Kingdom", "capital": "London"},
    ],
    "Africa": [
        {"country": "Egypt", "capital": "Cairo"},
        {"country": "Nigeria", "capital": "Abuja"},
        {"country": "South Africa", "capital": "Pretoria"},
        {"country": "Kenya", "capital": "Nairobi"},
        {"country": "Morocco", "capital": "Rabat"},
    ],
}


def build_tree(data):
    root = {"name": "Earth", "children": []}
    for continent, countries in data.items():
        continent_node = {"name": f"{continent} (Continent)", "children": []}
        for country_data in countries:
            country_node = {
                "name": f"{country_data['country']} (Country)",
                "children": [{"name": f"{country_data['capital']} (Capital)"}],
            }
            continent_node["children"].append(country_node)
        root["children"].append(continent_node)
    return root


if __name__ == "__main__":
    country_tree = build_tree(COUNTRY_DATA)
    print(tree(country_tree, children="children"))
