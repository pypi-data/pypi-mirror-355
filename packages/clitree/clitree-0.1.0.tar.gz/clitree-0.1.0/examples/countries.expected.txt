Earth
├── Europe (Continent)
│   ├── France (Country)
│   │   └── Paris (Capital)
│   ├── Germany (Country)
│   │   └── Berlin (Capital)
│   ├── Italy (Country)
│   │   └── Rome (Capital)
│   ├── Spain (Country)
│   │   └── Madrid (Capital)
│   └── United Kingdom (Country)
│       └── London (Capital)
└── Africa (Continent)
    ├── Egypt (Country)
    │   └── Cairo (Capital)
    ├── Nigeria (Country)
    │   └── Abuja (Capital)
    ├── South Africa (Country)
    │   └── Pretoria (Capital)
    ├── Kenya (Country)
    │   └── Nairobi (Capital)
    └── Morocco (Country)
        └── Rabat (Capital)
